File: codes/linked_list/copy_list_with_random_pointer.cpp
#include "../base.h"

// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if (head == NULL)
            return NULL;
        
        Node* dummy = head;

        // insert new nodes into the exited list 
        while (head) {
            Node* node = new Node(head->val);
            node->next = head->next;            
            head->next = node;
            head = node->next;
        }
        
        head = dummy;
        while (head && head->next) {
            if (head->random)
                head->next->random = head->random->next;
            head = head->next->next;
        }
        

        Node* newDummy = new Node(0);        
        Node* cur = newDummy;
        head = dummy;
        while (head) {
            cur->next = head->next;
            if (head->next)
                head->next = head->next->next;
            cur = cur->next;
            head = head->next;
        }
        
        return newDummy->next;        
    }
};

File: codes/linked_list/copy_list_with_random_pointer.py
from typing import Optional

# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        m = {}

        sh = head
        dummy = Node(0)
        cur = dummy
        while head:
            cur.next = Node(head.val)
            m[head] = cur.next
            cur = cur.next
            head = head.next

        head = sh
        while head:
            if head.random:
                m[head].random = m[head.random]
            head = head.next

        return dummy.next

class Solution_recursive:
    def __init__(self):
        self.visitedHash = {}

    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if head == None:
            return None

        if head in self.visitedHash:
            return self.visitedHash[head]

        node = Node(head.val, None, None)
        self.visitedHash[head] = node

        node.next = self.copyRandomList(head.next)
        node.random = self.copyRandomList(head.random)

        return node